File: HeraldTest-Flutter2.0/android/app/src/main/java/com/example/herald_flutter/SimpleKalmanFilter.java
package com.example.herald_flutter;

import java.lang.Math;

import androidx.annotation.Nullable;

// Simple 1-D Kalman Filter based on the code available at:
// https://github.com/denyssene/SimpleKalmanFilter
public class SimpleKalmanFilter {
    double errMeasure;
    double errEstimate;
    double q;
    double currentEstimate;
    double lastEstimate;
    double kalmanGain;

    public SimpleKalmanFilter(final double errMeasure, final double errEst, final double pNoise) {
        this.errMeasure = errMeasure;
        this.errEstimate = errEst;
        this.q = pNoise;
        this.currentEstimate = 0;
        this.lastEstimate = 0;
        this.kalmanGain = 0;
    }

    @Nullable
    public Double updateEstimate(Double measure) {
        if (measure == null) return null;
        kalmanGain = errEstimate / (errEstimate + errMeasure);
        currentEstimate = lastEstimate + kalmanGain * (measure - lastEstimate);
        errEstimate = (1.0 - kalmanGain) * errEstimate + Math.abs(lastEstimate - currentEstimate) * q;
        lastEstimate = currentEstimate;
        return currentEstimate;     
    }

    public void setMeasurementError(final double errMeasure) {
        this.errMeasure = errMeasure;
    }

    public void setEstimateError(final double errEst) {
        this.errEstimate = errEst;
    }
        
    public void setProcessNoise(final double pNoise) {
        this.q = pNoise;
    }

    public double getKalmanGain() {
        return this.kalmanGain;
    }
            
    public double getEstimateError() {
        return this.errEstimate;
    }        
}

File: ios/Runner/SimpleKalmanFilter.swift
//
//  SimpleKalmanFilter.swift
//  Runner
//
//  Created by Andrés Colubri on 8/9/22.
//

import Foundation

// Simple 1-D Kalman Filter based on the code available at:
// https://github.com/denyssene/SimpleKalmanFilter
public class SimpleKalmanFilter {
    public var errMeasure: Double
    public var errEstimate: Double
    public var q: Double
    public var currentEstimate: Double
    public var lastEstimate: Double
    public var kalmanGain: Double
    
    public init(_ errMeasure: Double, _ errEst: Double, _ pNoise: Double) {
        self.errMeasure = errMeasure
        self.errEstimate = errEst
        self.q = pNoise
        self.currentEstimate = 0
        self.lastEstimate = 0
        self.kalmanGain = 0
    }
    
    public func updateEstimate(_ measure: Double?) -> Double? {
        guard let measureValue = measure else {
            return nil
        }
        
        kalmanGain = errEstimate / (errEstimate + errMeasure)
        currentEstimate = lastEstimate + kalmanGain * (measureValue - lastEstimate)
        errEstimate = (1.0 - kalmanGain) * errEstimate + abs(lastEstimate - currentEstimate) * q
        lastEstimate = currentEstimate
        return currentEstimate
    }
    
    public func setMeasurementError(_ errMeasure: Double) {
        self.errMeasure = errMeasure
    }

    public func setEstimateError(_ errEst: Double) {
        self.errEstimate = errEst
    }
        
    public func setProcessNoise(_ pNoise: Double) {
        self.q = pNoise
    }

    public func getKalmanGain() -> Double {
        return self.kalmanGain
    }
            
    public func getEstimateError() -> Double {
        return self.errEstimate
    }
}
